cms/djangoapps/modulestore_migrator/api/read_api.py

```diff
@@ -0,0 +1,257 @@
+    """
+    return {
+        block_migration.source.key: _block_migration_result(block_migration)
+        for block_migration in models.ModulestoreBlockMigration.objects.filter(
+            overall_migration_id=migration_pk
+        ).select_related(
+            "source",
+            "target__learning_package",
+            # For building component key
+            "target__component__component_type",
+            # For building container key
+            "target__container__section",
+            "target__container__subsection",
+            "target__container__unit",
+            # For determining title and version
+            "change_log_record__new_version",
+        )
+    }
+
+
+def _migration(m: models.ModulestoreMigration) -> ModulestoreMigration:
+    """
+    Build a migration dataclass from the database row
+    """
+    return ModulestoreMigration(
+        pk=m.id,
+        source_key=m.source.key,
+        target_key=LibraryLocatorV2.from_string(m.target.key),
+        target_title=m.target.title,
+        target_collection_slug=(m.target_collection.key if m.target_collection else None),
+        target_collection_title=(m.target_collection.title if m.target_collection else None),
+        is_failed=m.is_failed,
+        task_uuid=m.task_status.uuid,
+    )
+
+
+def _block_migration_result(m: models.ModulestoreBlockMigration) -> ModulestoreBlockMigrationResult:
+    """
+    Build an instance of the migration result (successs/failure) dataclass from a database row
+    """
+    return (
+        _block_migration_success(
+            source_key=m.source.key,
+            target=m.target,
+            change_log_record=m.change_log_record,
+        )
+        if m.target
+        else ModulestoreBlockMigrationFailure(
+            source_key=m.source.key,
+            unsupported_reason=(m.unsupported_reason or ""),
+        )
+    )
+
+
+def _block_migration_success(
+    source_key: UsageKey,
+    target: PublishableEntity,
+    change_log_record: DraftChangeLogRecord | None,
+) -> ModulestoreBlockMigrationSuccess:
+    """
+    Build an instance of the migration success dataclass
+    """
+    target_library_key = LibraryLocatorV2.from_string(target.learning_package.key)
+    target_key: LibraryUsageLocatorV2 | LibraryContainerLocator
+    if hasattr(target, "component"):
+        target_key = library_component_usage_key(target_library_key, target.component)
+    elif hasattr(target, "container"):
+        target_key = library_container_locator(target_library_key, target.container)
+    else:
+        raise ValueError(f"Entity is neither a container nor component: {target}")
+    # We expect the block migration to have a DraftChangeLogRecord associated with it, which
+    # tells us the entity's version number and the title at the point immediately after the
+    # migration occured. However, the data model does not guarantee that the record exists.
+    # So, if the record is missing, we:
```

Comment:
Question: In what situation would the record not exist?

Reply:
Yeah, I think just having a clarifying comment in the code for how we might get into that state is fine.